### PR TITLE
[WIP] zfs-0.7.10 patchset

### PR DESCRIPTION
**Pull request by tonyhutter**
### Motivation and Context
Patchset for zfs-0.7.10

### How Has This Been Tested?
Run though buildbot

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by marcominetti** [comment]
Is anyone working on this? Can the linux kernel 4.18 support can be extracted in a quick release?

**Comment by tonyhutter** [comment]
I'm currently hitting an assert in one of the test cases:
```
[  294.081428] VERIFY(!RW_LOCK_HELD(&zv->zv_suspend_lock)) failed
[  294.081431] PANIC at zvol.c:1708:zvol_free()
[  294.081433] Showing stack for process 1994
[  294.081435] CPU: 2 PID: 1994 Comm: spl_system_task Tainted: P           OE     4.18.0-041800-generic #201808122131
[  294.081436] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.10.2-2.fc27 04/01/2014
[  294.081437] Call Trace:
[  294.081443]  dump_stack+0x63/0x83
[  294.081455]  spl_dumpstack+0x42/0x50 [spl]
[  294.081462]  spl_panic+0xc8/0x110 [spl]
[  294.081469]  ? spl_dumpstack+0x50/0x50 [spl]
[  294.081474]  ? stack_trace_call+0x61/0x70
[  294.081479]  ? ftrace_call+0x5/0x34
[  294.081480]  ? ftrace_call+0x5/0x34
[  294.081485]  ? spl_panic+0x5/0x110 [spl]
[  294.081566]  zvol_free+0x133/0x1d0 [zfs]
[  294.081572]  taskq_thread+0x2e5/0x5f0 [spl]
[  294.081575]  ? wake_up_q+0x80/0x80
[  294.081580]  kthread+0x120/0x140
[  294.081584]  ? taskq_thread_should_stop+0x70/0x70 [spl]
[  294.081585]  ? kthread_bind+0x40/0x40
[  294.081588]  ret_from_fork+0x35/0x40
```
I'm looking into it...

**Comment by loli10K** [comment]
I think there might be something wrong in `zvol_release()`, i'm attached to a local debug builder:

```
(gdb) bt
#0  __blkdev_put (bdev=0xffff8800d0709040, mode=<optimized out>, for_part=-797929168) at fs/block_dev.c:1544
#1  0xffffffff8122fc69 in blkdev_put (bdev=<optimized out>, mode=<optimized out>) at fs/block_dev.c:1607
#2  0xffffffff812f8028 in register_disk (disk=<optimized out>) at block/genhd.c:561
#3  add_disk (disk=0xffff8800bfbe5800) at block/genhd.c:620
#4  0xffffffffc06bb0e0 in zvol_create_minor_impl (name=0xffff8800bf855120 "testpool/vol") at /usr/src/zfs/module/zfs/zvol.c:1837
#5  0xffffffffc06bb27e in zvol_create_minors_impl (name=<optimized out>) at /usr/src/zfs/module/zfs/zvol.c:2066
#6  0xffffffffc06bb605 in zvol_task_cb (param=0xffff8800bf853a18) at /usr/src/zfs/module/zfs/zvol.c:2362
#7  0xffffffffc02a7a65 in taskq_thread (args=0xffff8800db345480) at /usr/src/spl/module/spl/spl-taskq.c:943
#8  0xffffffff8109ac4f in kthread (_create=0xffff880029444c00) at kernel/kthread.c:209
#9  0xffffffff815cc3f2 in ret_from_fork () at arch/x86/entry/entry_64.S:406
#10 0x0000000000000000 in ?? ()
(gdb) p disk->fops->release
$34 = (void (*)(struct gendisk *, fmode_t)) 0xffffffffc06ba300 <zvol_release>
(gdb) p ((zvol_state_t*)disk->private_data)->zv_suspend_lock 
$35 = {
  rw_rwlock = {
    count = 1, 
    wait_list = {
      next = 0xffff8800bf853a38, 
      prev = 0xffff8800bf853a38
    }, 
    wait_lock = {
      raw_lock = {
        val = {
          counter = 0
        }
      }
    }, 
    osq = {
      tail = {
        counter = 0
      }
    }, 
    owner = 0x0 <irq_stack_union>
  }
}
(gdb) p ((zvol_state_t*)disk->private_data)->zv_open_count 
$36 = 0
(gdb) 
```

We just called `zvol_release()` (`disk->fops->release`) and `zvol_state_t->zv_open_count` is now 0, but somehow we have 1 reader on the `zvol_state_t->zv_suspend_lock` (count = 1, owner = 0x0).

On master we initialize `boolean_t drop_suspend = B_TRUE` in `zvol_release()`, but in the zfs-0.7-release we have `boolean_t drop_suspend = B_FALSE`: it seems we might have taken not enough code from master (https://github.com/zfsonlinux/zfs/commit/5146d802b4e371cab1d6db79bea482c056be7bf2) in https://github.com/zfsonlinux/zfs/pull/7792/commits/9161ace4a02ea37b45f7ae613ea1d095a65d588d.

File: lib/libzfs/libzfs_sendrecv.c [hunk 1]
```diff
@@ -3592,7 +3592,7 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 		goto out;
 	}
 
-	if (top_zfs && *top_zfs == NULL)
+	if (top_zfs && (*top_zfs == NULL || strcmp(*top_zfs, name) == 0))
```
Comment (loli10K) [inline lib/libzfs/libzfs_sendrecv.c]: Fix from https://github.com/zfsonlinux/zfs/commit/3cbe89b12a55a0ff350d7434834f4f48227896a3 requires the following bits from the first encryption commit (https://github.com/zfsonlinux/zfs/commit/b52563034230b35f0562b6f40ad1a00f02bd9a05):

```diff
diff --git a/lib/libzfs/libzfs_sendrecv.c b/lib/libzfs/libzfs_sendrecv.c
index c5acd21..cadf16c 100644
--- a/lib/libzfs/libzfs_sendrecv.c
+++ b/lib/libzfs/libzfs_sendrecv.c
@@ -3577,6 +3577,7 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
                }
 
                newfs = B_TRUE;
+               *cp = '/';
        }
 
        if (flags->verbose) {
```

This should resolve "receive-o-x_props_override" test failure.

**Comment by loli10K** [comment]
And finally https://github.com/zfsonlinux/zfs/commit/f0ed6c744872ec6dc4838947ffc597f4d141864a requires a couple of functions (`get_debug_device`, `get_scsi_host`) in "blkdev.shlib" from https://github.com/zfsonlinux/zfs/commit/d3f2cd7e3b70679f127dd471ea6d37ece27463f2 to make the builders happy.

**Comment by kpande** [comment]
#7820 may be good to pull in here.

**Comment by tonyhutter** [comment]
@loli10K thanks, I added in those blkdev.shlib fixes along with your one-liner `receive-o-x_props_override` fix.
@kpande is there a patch out for #7820 yet?

**Comment by tonyhutter** [comment]
Released: https://github.com/zfsonlinux/zfs/releases/tag/zfs-0.7.10

**Pull request closed without merging**